### PR TITLE
MAILBOX-279: make the JPAMapperTests should work when enable MAILBOX and MESSAGE

### PR DESCRIPTION
**Pull request by quynhn**
which are:
- Split MessageMapperTest into 2 parts: one with attachment and the another without attachment which should be work with JPA
- Make JPAMapperTests should be fine when enable MAILBOX

### CODE REVIEW

**Comment by quynhn** [comment]
test this please

**Comment by quynhn** [comment]
test this please

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssert.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        assertThat(actual).extracting("name").containsAll(toNames(expecteds));
+        assertThat(actual).extracting("namespace").containsAll(toNamespaces(expecteds));
+    }
+
+    public void containAll(List<SimpleMailbox> expecteds) {
+        assertThat(actual).extracting("mailboxId").containsAll(toIds(expecteds));
+        assertThat(actual).extracting("user").containsAll(toUsers(expecteds));
+        assertThat(actual).extracting("name").containsAll(toNames(expecteds));
+        assertThat(actual).extracting("namespace").containsAll(toNamespaces(expecteds));
+    }
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssert.java]: With ContainsAll we can have extra entries. It's quitte weak.

Please provides containsOnly and containsExactly on you Assertions.

Moreover, Immagine : 

```
actlual = Immutablist.of(new Mailbox(id1, user1, namespace1, name1), new Mailbox(id2, user2, namespace2, name2));
expected = Immutablist.of(new Mailbox(id1, user1, namespace2, name2), new Mailbox(id2, user2, namespace1, name1));
```

Your contains methos will not fail but they should have...

What you need is to convert both mailbox lists into a common data type with your equals and hashcode. No need for it to be a Mailbox implementation. It can be an internal class. It would avoid mixing fields.

Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssert.java]: Where are the tests for this class ?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageWithAttachmentMapperTest.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+import org.junit.Assume;
+import org.junit.Rule;
+import org.junit.rules.ExpectedException;
+import org.xenei.junit.contract.Contract;
+import org.xenei.junit.contract.ContractTest;
+import org.xenei.junit.contract.IProducer;
+
+import com.google.common.collect.ImmutableList;
+
+//@Contract(MapperProvider.class)
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageWithAttachmentMapperTest.java]: We want to run it, no?

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMailboxMapper.java [hunk 3]
```diff
@@ -88,22 +93,31 @@ public void save(Mailbox mailbox) throws MailboxException {
     public Mailbox findMailboxByPath(MailboxPath mailboxPath) throws MailboxException, MailboxNotFoundException {
         try {
             if (mailboxPath.getUser() == null) {
-                return (Mailbox) getEntityManager().createNamedQuery("findMailboxByName").setParameter("nameParam", mailboxPath.getName()).setParameter("namespaceParam", mailboxPath.getNamespace()).getSingleResult();
+                return (Mailbox) getEntityManager().createNamedQuery("findMailboxByName")
```
Comment (quynhn) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMailboxMapper.java]: Done

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMailboxMapper.java [hunk 4]
```diff
@@ -88,22 +93,31 @@ public void save(Mailbox mailbox) throws MailboxException {
...
+                    .setParameter("nameParam", mailboxPath.getName())
+                    .setParameter("namespaceParam", mailboxPath.getNamespace())
+                    .getSingleResult();
             } else {
-                return (Mailbox) getEntityManager().createNamedQuery("findMailboxByNameWithUser").setParameter("nameParam", mailboxPath.getName()).setParameter("namespaceParam", mailboxPath.getNamespace()).setParameter("userParam", mailboxPath.getUser()).getSingleResult();
+                return (Mailbox) getEntityManager().createNamedQuery("findMailboxByNameWithUser")
```
Comment (quynhn) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMailboxMapper.java]: Done

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMailboxMapper.java [hunk 5]
```diff
@@ -88,22 +93,31 @@ public void save(Mailbox mailbox) throws MailboxException {
...
-        } 
+        }
     }
 
     @Override
     public Mailbox findMailboxById(MailboxId id) throws MailboxException, MailboxNotFoundException {
         JPAId mailboxId = (JPAId)id;
         try {
-            return (Mailbox) getEntityManager().createNamedQuery("findMailboxById").setParameter("idParam", mailboxId.getRawId()).getSingleResult();
+            return (Mailbox) getEntityManager().createNamedQuery("findMailboxById")
```
Comment (quynhn) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMailboxMapper.java]: Done

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/openjpa/AbstractJPAMailboxMessage.java [hunk 6]
```diff
@@ -228,6 +228,10 @@ public boolean equals(Object obj) {
             @ElementJoinColumn(name = "MAIL_UID", referencedColumnName = "MAIL_UID") })
     private List<JPAUserFlag> userFlags;
 
+    public AbstractJPAMailboxMessage() {
```
Comment (quynhn) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/openjpa/AbstractJPAMailboxMessage.java]: It has  a warning when I run the test. It said:
```
720  James  WARN   [main] openjpa.Enhance - The type "org.apache.james.mailbox.jpa.mail.model.openjpa.AbstractJPAMailboxMessage" must have a no-args constructor. Adding a public no-args constructor.
```
So I added that.

File: mailbox/lucene/src/test/java/org/apache/james/mailbox/lucene/search/LuceneMailboxMessageSearchIndexTest.java [hunk 7]
```diff
@@ -606,12 +606,16 @@ public void notOperatorShouldReverseMatching() throws Exception {
     }
     
     private final class SimpleMailbox implements Mailbox {
-        private final TestId id;
+        private TestId id;
```
Comment (quynhn) [inline mailbox/lucene/src/test/java/org/apache/james/mailbox/lucene/search/LuceneMailboxMessageSearchIndexTest.java]: Test's still ok so yes, I reverted the change

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/JPAMailbox.java [hunk 8]
```diff
@@ -79,45 +80,52 @@
...
      */
     public JPAId getMailboxId() {
         return JPAId.of(mailboxId);
     }
 
+    @Override
+    public void setMailboxId(MailboxId mailboxId) {
+        this.mailboxId = ((JPAId)mailboxId).getRawId();
+    }
+
```
Comment (rouazana) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/JPAMailbox.java]: This change seems to be related to an other commit

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssert.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+import java.util.List;
+
+import org.apache.james.mailbox.model.MailboxId;
+import org.apache.james.mailbox.store.mail.model.impl.SimpleMailbox;
+
+import com.google.common.base.Function;
+import com.google.common.collect.FluentIterable;
+
+public class ListMailboxAssert {
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssert.java]: Where (and why) do you need this Assert?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/SimpleMailbox.java [hunk 10]
```diff
@@ -138,7 +138,9 @@ public int hashCode() {
         final int PRIME = 31;
         int result = 1;
         result = PRIME * result + namespace.hashCode();
-        result = PRIME * result + user.hashCode();
+        if (user != null) {
```
Comment (quynhn) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/SimpleMailbox.java]: Done

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/MailboxEventDispatcherTest.java [hunk 11]
```diff
@@ -113,6 +114,10 @@ public MailboxACL getACL() {
         public void setACL(MailboxACL acl) {
```
Comment (quynhn) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/MailboxEventDispatcherTest.java]: Yes and Done

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssert.java (hunk 1 above)
Comment (quynhn) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssert.java]: It had wrong commit :( so it missed the changed file

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageWithAttachmentMapperTest.java (hunk 2 above)
Comment (quynhn) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageWithAttachmentMapperTest.java]: We want it, reverted the change

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/JPAMailbox.java (hunk 8 above)
Comment (quynhn) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/JPAMailbox.java]: Yes, it belongs to another commit, it will move it to the right commit

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssert.java (hunk 9 above)
Comment (quynhn) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssert.java]: Sorry, I missed the changed file which use this class

File: mailbox/lucene/src/test/java/org/apache/james/mailbox/lucene/search/LuceneMailboxMessageSearchIndexTest.java [hunk 12]
```diff
@@ -606,14 +606,13 @@ public void notOperatorShouldReverseMatching() throws Exception {
...
     private final class SimpleMailbox implements Mailbox {
-        private TestId id;
+        private final TestId id;
 
         public SimpleMailbox(long id) {
         	this.id = TestId.of(id);
         }
 
         public void setMailboxId(MailboxId id) {
-            this.id = (TestId)id;
```
Comment (rouazana) [inline mailbox/lucene/src/test/java/org/apache/james/mailbox/lucene/search/LuceneMailboxMessageSearchIndexTest.java]: Sorry, I'm note sure for this part of the fixup. Wha's the point of an empty setter?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssert.java [hunk 13]
```diff
@@ -27,48 +27,28 @@
 import org.apache.james.mailbox.store.mail.model.impl.SimpleMailbox;
 
 import com.google.common.base.Function;
+import com.google.common.base.Objects;
 import com.google.common.collect.FluentIterable;
 
 public class ListMailboxAssert {
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssert.java]: It is not possible to define a proper equal on Mailbox?
Else you could use isEqualToComparingOnlyGivenFields, no?

File: mailbox/lucene/src/test/java/org/apache/james/mailbox/lucene/search/LuceneMailboxMessageSearchIndexTest.java (hunk 12 above)
Comment (quynhn) [inline mailbox/lucene/src/test/java/org/apache/james/mailbox/lucene/search/LuceneMailboxMessageSearchIndexTest.java]: This is data for testing only and the attribute id is final so it cannot be changed, that's why I let the empty setter.

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssert.java (hunk 13 above)
Comment (quynhn) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssert.java]: We had "MailboxAssert" to compare Mailbox by its properties but in our cases, we have to compare the list of JPAMailbox and SimpleMailbox.
The method "isEqualToComparingOnlyGivenFields" is better to replace for MailboxAssert but I cannot do it because the Mailbox and SimpleMailbox do not have the same property names even. One has "id" and another has "mailboxId" so I think we need to refactor it. Do we have any other ways?

Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssert.java]: Thank you for the explantions. There would be other ways, but they would be more complicated.

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssert.java (hunk 1 above)
Comment (quynhn) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssert.java]: Done

**Comment by quynhn** [comment]
test this please

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssert.java [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    class InnerMailbox {
+        private final MailboxId id;
+        private final String user;
+        private final String name;
+        private final String namespace;
+
+        public InnerMailbox(MailboxId id, String user, String name, String namespace) {
+            super();
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssert.java]: super(); not needed

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssert.java [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+
+    public static ListMailboxAssert assertMailboxes(List<Mailbox> actual) {
+        return new ListMailboxAssert(actual);
+    }
+
+    public void containOnly(List<SimpleMailbox> expecteds) {
+        assertThat(mailboxtoInnerMailbox(actual)).containsOnlyElementsOf(simpleMailboxtoInnerMailbox(expecteds));
+    }
+
+    class InnerMailbox {
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssert.java]: private final class...

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MailboxMapperTest.java [hunk 16]
```diff
@@ -153,10 +125,14 @@ public void saveWithNullUserShouldPersistTheMailbox() throws MailboxException{
     public void listShouldRetrieveAllMailbox() throws MailboxException {
         saveAll();
         List<Mailbox> mailboxes = mailboxMapper.list();
-        assertThat(mailboxes).contains(benwaInboxMailbox, benwaWorkMailbox, benwaWorkTodoMailbox, benwaPersoMailbox, benwaWorkDoneMailbox, bobInboxMailbox, esnDevGroupInboxMailbox, esnDevGroupHublinMailbox,
-            esnDevGroupJamesMailbox, obmTeamGroupInboxMailbox, obmTeamGroupOPushMailbox, obmTeamGroupRoundCubeMailbox);
-    }
 
+        ListMailboxAssert.assertMailboxes(mailboxes)
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MailboxMapperTest.java]: You can static import ;-)

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMailboxMapper.java [hunk 17]
```diff
@@ -76,7 +77,11 @@ protected void commit() throws MailboxException {
     public void save(Mailbox mailbox) throws MailboxException {
         try {
             this.lastMailboxName = mailbox.getName();
-            getEntityManager().persist(mailbox);
+            JPAMailbox persistData = new JPAMailbox(mailbox);
```
Comment (quynhn) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMailboxMapper.java]: Done

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMailboxMapper.java [hunk 18]
```diff
@@ -131,9 +146,16 @@ public void delete(Mailbox mailbox) throws MailboxException {
...
         try {
             if (path.getUser() == null) {
-                return getEntityManager().createNamedQuery("findMailboxWithNameLike").setParameter("nameParam", SQL_WILDCARD_CHAR + path.getName() + SQL_WILDCARD_CHAR).setParameter("namespaceParam", path.getNamespace()).getResultList();
+                return getEntityManager().createNamedQuery("findMailboxWithNameLike")
+                    .setParameter("nameParam", path.getName())
+                    .setParameter("namespaceParam", path.getNamespace())
+                    .getResultList();
             } else {
-                return getEntityManager().createNamedQuery("findMailboxWithNameLikeWithUser").setParameter("nameParam", SQL_WILDCARD_CHAR + path.getName() + SQL_WILDCARD_CHAR).setParameter("namespaceParam", path.getNamespace()).setParameter("userParam", path.getUser()).getResultList();
+                return getEntityManager().createNamedQuery("findMailboxWithNameLikeWithUser")
```
Comment (quynhn) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMailboxMapper.java]: Done

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssert.java (hunk 15 above)
Comment (quynhn) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssert.java]: Done

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssert.java (hunk 14 above)
Comment (quynhn) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssert.java]: Done

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssert.java [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+
+    private ListMailboxAssert(List<Mailbox> actual) {
+        this.actual = actual;
+    }
+
+    public static ListMailboxAssert assertMailboxes(List<Mailbox> actual) {
+        return new ListMailboxAssert(actual);
+    }
+
+    public void containOnly(List<Mailbox> expecteds) {
```
Comment (mbaechler) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssert.java]: this one should be containsOnlyElementsOf and you should define a containsOnly with takes a vararg to make calling easier

Comment (quynhn) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssert.java]: Done

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssertTest.java [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    @Test
+    public void assertListMailboxShouldWork() throws Exception {
+        assertMailboxes(actualMailbox).containOnly(createMailbox(NAMESPACE, USER, NAME, UID_VALIDITY), 
+            createMailbox(OTHER_NAMESPACE, USER, NAME, UID_VALIDITY));
+    }
+    
+    private Mailbox createMailbox(final String namespace, final String user, final String name, final long uid_validity) {
+        return new Mailbox() {
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMailboxAssertTest.java]: Use SimpleMailbox ? Or Mockito ?

**Comment by quynhn** [comment]
test this please

**Comment by quynhn** [comment]
test this please

**Comment by quynhn** [comment]
test this please
